src/datasource.test.ts

```diff
@@ -0,0 +1,177 @@
+import { DataQueryRequest, DataSourceInstanceSettings, dateTime } from '@grafana/data';
+import { BackendSrv, setBackendSrv, setTemplateSrv } from '@grafana/runtime';
+import { TemplateSrv } from 'tests/template_srv';
+import { AthenaDataSourceOptions, AthenaQuery, FormatOptions } from 'types';
+
+import { DataSource } from './datasource';
+
+interface TestContext {
+  instanceSettings: DataSourceInstanceSettings<AthenaDataSourceOptions>;
+  ds: DataSource;
+}
+
+describe('AthenaDatasource', () => {
+  const ctx: TestContext = {} as TestContext;
+  const backendSrv = {
+    get: (url: string, params?: any, requestId?: string) => {},
+    post: (url: string, data?: any) => {},
+  };
+  const templateSrv = new TemplateSrv();
+
+  const defaultQuery: AthenaQuery = {
+    connectionArgs: {
+      region: 'defaultRegion',
+      catalog: 'defaultCatalog',
+      database: 'defaultDatabase',
+    },
+    table: 'defaultTable',
+    column: 'defaultColumn',
+    refId: 'testQuery',
+    rawSQL: '',
+    format: FormatOptions.TimeSeries,
+  };
+  const start = dateTime(new Date());
+  const timeRange = { from: start, to: start.add(3600 * 1000, 'millisecond') };
+  const rawTimeRange = { from: timeRange.from.toString(), to: timeRange.to.toString() };
+  const queryRequest: DataQueryRequest<AthenaQuery> = {
+    ...defaultQuery,
+    range: { ...timeRange, raw: rawTimeRange },
+    requestId: 'testRequest',
+    interval: '1000',
+    intervalMs: 1000,
+    scopedVars: {},
+    startTime: start.valueOf(),
+    timezone: 'TZ',
+    targets: [defaultQuery],
+    app: 'athena',
+  };
+
+  beforeEach(() => {
+    jest.clearAllMocks();
+
+    ctx.instanceSettings = {
+      name: 'testAthena',
+      jsonData: { defaultRegion: 'testRegion', catalog: 'testCatalog', database: 'testDatabase' },
+    } as unknown as DataSourceInstanceSettings<AthenaDataSourceOptions>;
+    ctx.ds = new DataSource(ctx.instanceSettings);
+    setBackendSrv(backendSrv as BackendSrv);
+    setTemplateSrv(templateSrv);
+    jest.spyOn(templateSrv, 'replace');
+  });
+
+  describe('When performing getRegions', () => {
+    it('should return a list of regions', async () => {
+      const response = setupRegionsResponse();
+      backendSrv.get = jest.fn(() => Promise.resolve(response));
+      const regionsResponse = await ctx.ds.getRegions();
+
+      expect(regionsResponse).toHaveLength(response.length);
+      expect(regionsResponse).toBe(response);
+    });
+  });
+
+  describe('When performing getCatalogs', () => {
+    it('should return a list of catalogs', async () => {
+      const response = setupCatalogsResponse();
+      backendSrv.post = jest.fn(() => Promise.resolve(response));
+      const catalogsResponse = await ctx.ds.getCatalogs(defaultQuery);
+
+      expect(catalogsResponse).toHaveLength(response.length);
+      expect(catalogsResponse).toBe(response);
+      expect(templateSrv.replace).toHaveBeenCalledWith(defaultQuery.connectionArgs.region);
```

Comment:
nit: rather than checking that replace has been called, I would:
 - Create a new test that in the request it includes a "variable" (i.e. `{database: "$foo"}`)
 - Mock replace with a dummy result: `replace = jest.fn().mockImplementation((f: string) => f.replace("$foo", "bar"));`
 - Check that the response contains `bar` not `$foo`

Reply:
I prefer it since it rely a bit less in the implementation (even though you still need to mock an internal method) and more in the expected output.